### ISSUE TITLE
Ignore changes by performing provider read

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi#12346]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Provide the ability for users to specify properties which should not be modified from their current deployed state.

This is similar to the current ignoreChanges functionality but instead of reading the value from state it would perform a read via the provider for the resource then override the specified property (or properties) from the retrieved values. 

The resulting behaviour would set the property from the property specified in the Pulumi program. On subsequent updates, any manual modifications to the specified properties would be maintained, rather than reverting the property back to the value in the program. 

### Use Case 1 - Externally Managed Fields (e.g. Load Ballancer)

- On the first deployment of the load ballancer, an initial desired instance count can be set (e.g. 1).
- Once deployed, there may be an interactive process which modifies the desired instance count dynamically.
- On subsequent deployments, we want to avoid changing this desired count - but want to maintain the current value in the cloud service.

This is demonstrated in AWSx Classic where we use a CloudFormation template internally to achieve this behaviour: https://github.com/pulumi/pulumi-awsx/blob/95cf49fd4108cf7fdb5cbdf883c1f661dadb2b04/awsx-classic/autoscaling/autoscaling.ts#L251-L263

### Use Case 2 - Nested Resources

In azure-native, the Virtual Network Resource can have subnets created either via a property of the VNet, or as stand-alone Subnet resources. However, any standalone resources which are created will then appear in the `subnets` property of the VNet. Subsequent updates to the VNet will then cause stand-alone subnets to be deleted because they are not in the state for that property, so doing a PUT of the VNet without the subnets triggers a deletion of the subnets. 

Having an ignore changes which reads from the cloud provider rather than the state would allow users to quickly work-around this issue by reading the current state of the subnets property for use in the next deployment.

Reference: https://github.com/pulumi/pulumi-azure-native/issues/611

### Use Case 3 - Unintentionally exposed field (e.g. ETags)

Some cloud providers may use ETags as a mechanism to avoid race conditions between update operations. This is seen in GCP (and is handled specially by the provider). This would give a user-controllable "escape hatch" to avoid being blocked by other similar behaviours.

### Use Case 4 - Perma-diff

Programs can sometimes get perminant diffs due to subtle bugs in providers or a cloud API. For example a field might have its case changed on writing, so if the program's input doesn't match it will appear as an update. 

### Implementation Considerations

This could be an adjustment in behaviour of the existing `ignoreChanges` resource option, or could live as a new second resource option. 

If implemented as a change in behaviour to the existing `ignoreChange` option, we would want to initially put this behind a feature switch for users to opt-in to while we understand if there's possible issues or genuine need to use Pulumi state instead of the provider-read state. The default behaviour could be changed in a future major version change.

Rough outline of the engine implementation:

1. If the resource is using a our "new ignore changes" option, add a read operation to the plan before the update.
2. When performing the update, merge the inputs with the data from the read operation.
3. Persist the updated state from the result of the upate operation.

This is equivilent to using an invoke within the user's program to get the current value from the cloud, or falling back to a default.

An alternative approach could be to read the resource and immediately update the state of the specified properties to ignore and then perform the update steps as usual. This would be more equivelent to always performing a refresh (on selected resource properties) before the update.

**Comment by justinvp** [comment]
We'll be discussing on Thursday

**Comment by danielrbradley** [comment]
Summary from internal conversation:

This will be redundant if we move to a default of refreshing before every operation - which would be for consideration at the next major version. Also mentioned starting to introduce refreshes by default in a "what-if" mechanism - so we just report to the user where the refresh would have changed inputs, but don't actively reset them again.

Adding this mechanism would require waiting on additional calls to the cloud during the preview (which is currently a synchronous operation) impacting performance of that step and potentially adding complexity. The performance hit would not be worse than doing refresh before preview.

This could be implemented now on a per-provider basis, but would require being implemented independently by each provider and implementation details could vary. If implementing this mechanism in the engine, then we could also support ignoring properties by default via the schema as well as arbitary user-specified properties.

**Comment by segevfiner** [comment]
Not exactly. As the value in the source would still not match with the state. So it will still result in reverting to the source value when anything else changes even if the state has the latest value.

I think this behavior can be emulated by using a provider method to get the existing resource value (or a default) and using that in the source along with `ignoreChanges`.

Without a refresh the value in the state and source will differ but `ignoreChanges` will ignore it, and when something else changes, it will apply the same value that is already on the actual resource rather than reverting.

With a refresh, the value in the source and state will match, so `ignoreChanges` is probably not needed, but you still need to have the source point to the latest value. Which is what is currently missing as an option and you will need to emulate as I described.

**Comment by segevfiner** [comment]
The problem is that you need the actual name of the resource for invoking such a method, and there isn't an obvious way to get the current resource name from the state inside the Pulumi program. Also, there isn't a way to handle errors on a `pulumi.Output` so I think you must use the `Promise` version of the provider method.

**Comment by barakcoh** [comment]
We have a similar use-case, using Cognito and external IdPs. We'd like to configure manually via the AWS Console and have no need for them in the Pulumi program, as it's per-customer configuration, only needed in the production environment, and the support team doing the configuration has no access to the Pulumi code.

The problem is, even with `ignoreChanges`, any change to the `aws.cognito.UserPoolClient` will reset the `supportedIdentityProviders` (that has the state value of `[]`) effectively revoking access to all external IdPs. 

Is there a way around this other than the suggestion made by the OP?

**Comment by rcoundon** [comment]
> We have a similar use-case, using Cognito and external IdPs. We'd like to configure manually via the AWS Console and have no need for them in the Pulumi program, as it's per-customer configuration, only needed in the production environment, and the support team doing the configuration has no access to the Pulumi code.
> 
> The problem is, even with `ignoreChanges`, any change to the `aws.cognito.UserPoolClient` will reset the `supportedIdentityProviders` (that has the state value of `[]`) effectively revoking access to all external IdPs.
> 
> Is there a way around this other than the suggestion made by the OP?

I'm wondering if you found a workaround for this? I'm experiencing a similar but even more fundamental issue where an existing User Pool Client is deleted whenever it's imported

**Comment by Zaid-Ajaj** [comment]
Hi there @rcoundon @barakcoh, the problems you are facing, do they still persist? I would love to hear about them, especially with respect to import because we want to fix everything related to that. Please consider opening new issues with more details so that we can have a good look at these issues. Alternatively, feel free to contact me via zaid@pulumi.com to discuss this further. In the meantime, I will be closing this issue as won't-fix and `pulumi refresh --run-program` does what OP wanted (performing a READ before update)
